### ISSUE TITLE
Incompatibility between Rack::Session and ActionDispatch::Request::Session

### ISSUE THREAD

**Issue by pboling** [rails/rails#15843]
It is preventing integration testing in test-unit of my **Grape** app mounted in my Rails 4.1.1 routes.rb file.

Here's a simplified example: https://github.com/pboling/x-cascade_header_rails/tree/e34a94ba76f9350aed13486f282e99a98e2d9821

(Never mind the name of the repo... I was trying to understand the X-Cascade Header.)

Repro after checking out the repo, bundle install, rake db:migrate:

```
bin/rake test
...
  1) Failure:
XCascadeGrapeTest#test_/api/auth/login_invalid_credentials_is_401 [/Users/pboling/Documents/src/my/x-cascade_header_rails/test/integration/x_cascade_grape_test.rb:10]:
{"error":"undefined method `each' for #\u003cActionDispatch::Request::Session:0x103aebb00 not yet loaded\u003e"}.
Expected: 401
  Actual: 500
```

1. Here's another of the **exact** same error from **Sinatra** mounted in rails:
https://gist.github.com/toolmantim/9597022

2. Here is another report of the same error from a Grape app mounted in Rails 4.0.2:
http://stackoverflow.com/questions/22439361/rspec-testing-api-with-rack-protection

**Comment by joost** [comment]
Trying to mount a Rack app in Rails I got the following Rack::Lint::LintError
 (Lint assertion):

    session #<ActionDispatch::Request::Session:0x7f8466bfa3c8 not yet loaded> must respond to store and []=

This is because Rack::Lint expects session to respond to `store`. You could try monkey patching with something like:

    class ActionDispatch::Request::Session
      def store(name, value)
        self[name] = value
      end
    end

**Comment by pboling** [comment]
Any chance this could be addressed by the Rails 5 Gods?

It seems like the ideal scenario would be for the `ActionDispatch::Request::Session` to be more like an enhancement of the Rack Session, similar to how HashWithIndifferentAccess tries to play nicely with Hash.

:bump:

**Comment by kitop** [comment]
Just updated a Rails 3.2 app to Rails 4 and hit this same exact issue.

Is there any other solution besides monkey-patching?

+1 to @pboling suggestion of making it play nice with `Rack::Session` ala `HashWithIndifferentAccess` and `Hash`

**Comment by pboling** [comment]
@kitop I asked [rails core about this issue on twitter](https://twitter.com/galtzo/status/578333681619730432) and they responded.  It seems that there interest in addressing this in the upcoming Rails 5.

**Comment by abhinandankothari** [comment]
Just Created new app with Rails 4.2 and got this error

**Comment by mperham** [comment]
Warning: use at your own risk.  Monkeypatching is never a great idea, especially a private method.  Put this in `config/initializers/session_store.rb`:

```ruby
# Monkeypatch necessary due to https://github.com/rails/rails/issues/15843
require 'rack/session/abstract/id'
class Rack::Session::Abstract::SessionHash
  private
  def stringify_keys(other)
    hash = {}
    other = other.to_hash unless other.is_a?(Hash) # hack hack hack
    other.each do |key, value|
      hash[key.to_s] = value
    end
    hash
  end
end
```

**Comment by LouisSayers** [comment]
Another temporary solution that you might find helpful:

# Gemfile 
gem 'mocha'

# some_test.rb
def add_session_value(session_id, session_value)
    ActionDispatch::Request::Session.
      any_instance.stubs(:[]).with(session_id).
      returns(session_value)
  end

**Comment by maclover7** [comment]
@tenderlove might have some ideas about this...

**Comment by mperham** [comment]
Here's the monkeypatch that Sidekiq uses to work with Rails today:

https://github.com/mperham/sidekiq/blob/master/lib/sidekiq/web.rb#L266

**Comment by maclover7** [comment]
@mperham Would both the `each` and `stringify_keys` methods need to be added? Also, how would we go about testing this?

**Comment by mperham** [comment]
I don't know.

**Comment by maclover7** [comment]
Opened PR #24820 to try and solve this :)

**Comment by maclover7** [comment]
Fixed via #24820.
